### Задача 1
Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Составьте закон распределения числа правильных ответов при простом угадывании. Найдите M(X), D(X).

In [19]:
# Количество вопросов - 5
# Вариантов ответов - 4
# Правильных вариантов - 1

In [20]:
# Вероятность ответить правильно или неправильно в каждом вопросе не зависит
# от ответов на другие вопросы и эта вероятность всегда одинакова.
# Можно применить формулу Бернулли

In [3]:
from math import factorial

In [4]:
def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

In [28]:
# Вероятность ответить неправильно на все 5 вопросов
k0 = combinations(5, 0)*(1/4)**0*(1-1/4)**(5-0)
k0

0.2373046875

In [22]:
# Вероятность ответить правильно на 1 вопрос
k1 = combinations(5, 1)*(1/4)**1*(1-1/4)**(5-1)
k1

0.3955078125

In [23]:
# Вероятность ответить правильно на 2 вопроса
k2 = combinations(5, 2)*(1/4)**2*(1-1/4)**(5-2)
k2

0.263671875

In [25]:
# Вероятность ответить правильно на 3 вопроса
k3 = combinations(5, 3)*(1/4)**3*(1-1/4)**(5-3)
k3

0.087890625

In [26]:
# Вероятность ответить правильно на 4 вопроса
k4 = combinations(5, 4)*(1/4)**4*(1-1/4)**(5-4)
k4

0.0146484375

In [27]:
# Вероятность ответить правильно на все 5 вопросов
k5 = combinations(5, 5)*(1/4)**5*(1-1/4)**(5-5)
k5

0.0009765625

In [29]:
# Проверяем рассчеты вероятностей для правильных ответов
total = k0 + k1 + k2 + k3 + k4 + k5
total

1.0

In [49]:
import numpy as np

In [50]:
# Массив значений
x_values = np.arange(0, 6)
x_values

array([0, 1, 2, 3, 4, 5])

In [51]:
# Массив вероятностей
x_prob = np.array([k0, k1, k2, k3, k4, k5])
x_prob

array([0.23730469, 0.39550781, 0.26367188, 0.08789062, 0.01464844,
       0.00097656])

In [52]:
# Находим матожидание m
m = x_values.dot(x_prob)
m

1.25

In [53]:
# Находим дисперсию d
y_values = x_values - m
y_values

array([-1.25, -0.25,  0.75,  1.75,  2.75,  3.75])

In [54]:
z_values = y_values ** 2
z_values

array([ 1.5625,  0.0625,  0.5625,  3.0625,  7.5625, 14.0625])

In [55]:
d = z_values.dot(x_prob)
d

0.9375

In [56]:
# Для понимания разброса значений вправо и влево
# от матожидания находим среднеквадратическое отклонение s
s = d**0.5
s

0.9682458365518543

### Задача 2
Вероятность поймать рыбу при одном забрасывании удочки равна 0.01. Какова вероятность того, что число пойманных рыб при 200 забросах удочки будет не менее 5 и не более 10?

In [57]:
# В данной задаче достаточно большое число испытаний, также есть интенсивность
# т.о. она подходит под распределение Пуассона

In [70]:
a = 200*0.01
a

2.0

In [59]:
# k имеет диапазон значений 5 >= k <= 10
# Поэтому нам необходимо найти суммо вероятностей исходов данного диапазона

In [699]:
def poisson_proba(k, a):
    return (a ** k) * (np.exp(-a)) / np.math.factorial(k)

In [64]:
# Вероятность составит
sum(poisson_proba(k=i, a=a) for i in range(5,11))

0.052644709119342686

### Задача 3
В условиях предыдущей задачи сколько забросов удочки нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 10 раз будет поймана рыба?
Подсказка. Как посчитать P(X >= k) для величины X, имеющей биномиальное распределение? Здесь предстоит немножко покодить.

#### Вариант 1

In [692]:
# Вероятность должна быть = 0.9
# Поймать нужно >= 10 рыб

p = 0
k = 10
n = 0
spam = 1

In [693]:
# Находим число забрасываний путем их перебора до получения нужной вероятности
# от обратного, т.е. кол-во забрасываний, чтобы вытащить не более 10 рыб с
# с вероятностью не более 0.1

while spam != 0:
    n += 1
    a = n * 0.01
    p = sum(poisson_proba(k=k, a=a) for i in range(0,11))
    if p > 0.1:
        spam = 0
        n -= 1
        a = n * 0.01
        p = sum(poisson_proba(k=k, a=a) for i in range(0,11))
1- p, n, a

(0.9000639731696665, 439, 4.39)

In [694]:
# Проверяем
n = 439
a = n * 0.01
k = 10
1 - sum(poisson_proba(k=k, a=a) for i in range(0,11))

0.9000639731696665

По закону распределения Пуассона, чтоыбы поймать не менее 10 рыб с вероятностью 0.9 необходимо 439 забрасываний

#### Вариант 2

Пробуем найти по формуле Бернули

$$P_n(k>=10) = 0.9$$

Найдем от обратного, т.е. 1 - вероятность не поймать ни одной рыбы:
$$1 - P_n(k>=0) = 0.9$$
$$1 - C_n^0*0.01^0(1-0,01)^n = 0.9$$
$$1 - \frac{n!}{0!(n-0)!} \cdot 0.01^0 \cdot 0.99^n = 0.9$$
$$1 - \frac{n!}{n!} \cdot 1 \cdot 0.99^n = 0.9$$
$$1 - 0.99^n = 0.9$$
$$0.99^n = 0.1$$

In [707]:
n = 0
p = 1

In [708]:
while p > 0.1:
    n += 1
    p = 0.99**n
1 - p, n

(0.9008951844811254, 230)

По биномиальному закону распределения, чтоыбы поймать не менее 10 рыб с вероятностью 0.9 необходимо 230 забрасываний